{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "mkdir -p script t result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "download/PPMISI10874.diploidSV.vcf.gz\n",
      "download/PPMISI12224.diploidSV.vcf.gz\n",
      "download/PPMISI12499.diploidSV.vcf.gz\n",
      "download/PPMISI12593.diploidSV.vcf.gz\n",
      "download/PPMISI13039.diploidSV.vcf.gz\n",
      "download/PPMISI13424.diploidSV.vcf.gz\n",
      "download/PPMISI14281.diploidSV.vcf.gz\n",
      "download/PPMISI14331.diploidSV.vcf.gz\n",
      "download/PPMISI14426.diploidSV.vcf.gz\n",
      "download/PPMISI15761.diploidSV.vcf.gz\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "ls download/* |  head"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Derive SVs from Individual diploid file\n",
    "t = \"\"\"\n",
    "library(data.table);library(dplyr)\n",
    "# Paramaters\n",
    "fnames = list.files(\"manta/\")\n",
    "# filename = \"manta/PPMISI10874.diploidSV.vcf.gz\"\n",
    "# folder = tstrsplit(filename, \"/\") %>% .[[1]]\n",
    "pdgenes = fread('../../../tool/PDgenes/PD_GENES_2017_v1.csv', skip = 1) %>% filter(complete.cases(.)) %>% \n",
    "  rename(GENE = V8) %>% \n",
    "  mutate(chr = paste(\"chr\", chromosome, sep= \"\")) %>% \n",
    "  select(GENE, chr, start, end)\n",
    "\n",
    "for (i in 1:length(fnames)){\n",
    "  filename = paste(\"manta\", fnames[i], sep = \"/\") \n",
    "  PATNO = tstrsplit(filename, \"PPMISI\") %>% .[[2]] %>% tstrsplit(., \"\\\\\\\\.\") %>% .[[1]]\n",
    "  output = paste(\"t/\", \"PPMISI\", PATNO, sep = \"\")\n",
    "  \n",
    "  # Read diploid model outputs\n",
    "  dip = fread(cmd = paste(\"zless\", filename, \"| grep -v '##'\"))\n",
    "  dip = dip %>% filter(FILTER==\"PASS\")\n",
    "  dip = dip %>% filter(`#CHROM` %in% paste(\"chr\", c(1:22, \"X\"), sep = \"\")) %>%\n",
    "    rename(chr = `#CHROM`)\n",
    "  \n",
    "  # Filter SVs by their positions\n",
    "  j = inner_join(dip, pdgenes, by = \"chr\") %>% \n",
    "    mutate(PATNO = as.numeric(PATNO))\n",
    "  \n",
    "  f10K = j %>% \n",
    "    filter(POS> start - 10000 & POS < end + 10000) %>% # 10K margin\n",
    "    fwrite(., paste(output, '10K.txt', sep = \"_\"))\n",
    "  \n",
    "  f1K = j %>% \n",
    "    filter(POS> start - 1000 & POS < end + 1000) %>% # 1K margin\n",
    "    fwrite(., paste(output, '1K.txt', sep = \"_\"))\n",
    "}\n",
    "\"\"\"\n",
    "with open('script/deriveMantaOutpus.R', 'w') as f:\n",
    "    f.write(t)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Populate derived file\n",
    "t = \"\"\"\n",
    "library(data.table);library(dplyr)\n",
    "KEY = \"t/*_1K.txt\"\n",
    "tempfunc = function(K){\n",
    "  t = fread(cmd = paste(\"awk 'NR==1{print;next}FNR>1{print}'\", K))\n",
    "  names(t)[10]=\"GENOTYPE\"\n",
    "  dx = fread(\"../../../PDcohorts/PPMI/out181018/DEMOG_DIAG.csv\")\n",
    "  tab = semi_join(dx, t, by = \"PATNO\") %>% with(table(RECRUIT)) %>% data.frame(.) \n",
    "  j = inner_join(t, dx, by = \"PATNO\") %>% inner_join(., tab, by=\"RECRUIT\")\n",
    "  j %>% with(table(GENE, RECRUIT))\n",
    "  dip = j %>% \n",
    "    tidyr::separate(\"GENOTYPE\", c(\"GENOTYPE\", rep(NA, 5)), sep=\":\") %>% \n",
    "    mutate(DOSE = case_when(\n",
    "      GENOTYPE==\"1/1\" ~ 2, \n",
    "      GENOTYPE==\"1/0\" ~ 1,\n",
    "      GENOTYPE==\"0/1\" ~ 1,\n",
    "      GENOTYPE==\"0/0\" ~ 0,\n",
    "      TRUE ~ 99))\n",
    "  t = dip %>% select(ID) %>%\n",
    "    tidyr::separate(ID, paste(\"V\", 1:8, sep=\"\"), sep=\":\", fill = 'right')\n",
    "  dip$TYPE = ifelse(is.na(as.numeric(t$V2)), paste(t$V1, t$V2, sep = \":\"), t$V1)\n",
    "  \n",
    "  # Get the SV length\n",
    "  t = dip %>% select(INFO) %>%\n",
    "    tidyr::separate(INFO, paste(\"V\", 1:10, sep=\"\"), sep=\";\", fill = 'right')\n",
    "  getLEN = function(x){\n",
    "    v = t[x,] %>% t %>% as.vector\n",
    "    j = grep(\"SVLEN\\\\\\\\=\", v)\n",
    "    if(length(j)==1){\n",
    "      g = v[j]\n",
    "      SVLEN = tstrsplit(g, \"\\\\\\\\=\")[[2]][1] %>% as.numeric\n",
    "    }else{SVLEN=NA}\n",
    "    return(SVLEN)\n",
    "  }\n",
    "  dip$SVLEN = lapply(1:nrow(t), getLEN) %>% unlist\n",
    "  out = sub(\"t/\\\\\\\\*\", \"result/SV\", K) %>% sub(\"txt\", \"csv\", .) # Name for the output file\n",
    "  fwrite(dip, out, row.names = F)\n",
    "}\n",
    "tempfunc(KEY)\n",
    "tempfunc(\"t/*_10K.txt\")\n",
    "\"\"\"\n",
    "with open (\"script/populateDerivedSVs.R\", \"w\") as f:\n",
    "    f.write(t)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get table do burden test and explore SNCA\n",
    "t = \"\"\"\n",
    "library(data.table);library(dplyr)\n",
    "fname = \"result/SV_1K.csv\"\n",
    "dip = fread(fname)\n",
    "# Doses\n",
    "dip %>% with(table(GENE, DOSE)) %>% addmargins(., 2)\n",
    "\n",
    "## Burden table\n",
    "dft = dip %>% \n",
    "  # filter(TYPE!=\"MantaBND\") %>% \n",
    "  group_by(PATNO, GENE) %>% summarise_at(vars(\"DOSE\"), funs(sum(., na.rm = T))) %>% \n",
    "  tidyr::spread(key = GENE, value = DOSE)\n",
    "dft[is.na(dft)]=0\n",
    "dft = inner_join(dft, dx, by = \"PATNO\")\n",
    "dft %>% CreateContTable(vars = names(dft)[2:16], strata = \"RECRUIT\", data = .) %>% print(., explain = F, test = F)\n",
    "# Burden test\n",
    "## raw p\n",
    "rp = dft %>% filter(RECRUIT %in% c(\"HC\", \"PD\")) %>% \n",
    "  CreateContTable(vars = names(dft)[2:16], strata = \"RECRUIT\", data = .) %>% print(., explain = F) %>% \n",
    "  data.frame() %>% .[,1:3]\n",
    "## adj-p\n",
    "pca = fread(\"../WGS/clean_NONGENT/3a.eigenvec\") %>% mutate(PATNO = sub(\"PPMISI\",\"\", IID) %>% as.numeric)\n",
    "dfts = dft %>% filter(RECRUIT %in% c(\"HC\", \"PD\")) %>% inner_join(., pca, by =\"PATNO\") \n",
    "rp$`adj-p`=NA\n",
    "for (i in 2:nrow(rp)){\n",
    "  gene = rownames(rp)[i]\n",
    "  model = paste(\"I(RECRUIT=='PD') ~\", gene, \"+PC1+PC2+PC3+PC4+PC5+PC6+PC7+PC8+PC9+PC10\")\n",
    "  tdf = lm(model, data = dfts) %>% summary %>% coef\n",
    "  ans = try(tdf[gene, 4], silent = T)\n",
    "  if(class(ans)==\"try-error\"){rp$`adj-p`[i]=NA}else{rp$`adj-p`[i] = round(ans, 3)}\n",
    "}\n",
    "print(rp)\n",
    "\n",
    "## Test without MantaBND\n",
    "dft = dip %>% \n",
    "  filter(TYPE!=\"MantaBND\") %>%\n",
    "  group_by(PATNO, GENE) %>% summarise_at(vars(\"DOSE\"), funs(sum(., na.rm = T))) %>% \n",
    "  tidyr::spread(key = GENE, value = DOSE)\n",
    "dft[is.na(dft)]=0\n",
    "dft = inner_join(dft, dx, by = \"PATNO\")\n",
    "## raw p\n",
    "rp = dft %>% filter(RECRUIT %in% c(\"HC\", \"PD\")) %>% \n",
    "  CreateContTable(vars = names(dft)[2:15], strata = \"RECRUIT\", data = .) %>% print(., explain = F) %>% \n",
    "  data.frame() %>% .[,1:3]\n",
    "## adj-p\n",
    "dfts = dft %>% filter(RECRUIT %in% c(\"HC\", \"PD\")) %>% inner_join(., pca, by =\"PATNO\") \n",
    "rp$`adj-p`=NA\n",
    "for (i in 2:nrow(rp)){\n",
    "  gene = rownames(rp)[i]\n",
    "  model = paste(\"I(RECRUIT=='PD') ~\", gene, \"+PC1+PC2+PC3+PC4+PC5+PC6+PC7+PC8+PC9+PC10\")\n",
    "  tdf = lm(model, data = dfts) %>% summary %>% coef\n",
    "  ans = try(tdf[gene, 4], silent = T)\n",
    "  if(class(ans)==\"try-error\"){rp$`adj-p`[i]=NA}else{rp$`adj-p`[i] = round(ans, 3)}\n",
    "}\n",
    "print(rp)\n",
    "\n",
    "\n",
    "## SNCA\n",
    "snca = dip %>% filter(RECRUIT %in% c(\"PD\", \"HC\")) %>% \n",
    "  filter(GENE==\"SNCA\") \n",
    "t = snca %>% \n",
    "  group_by(POS, REF, ALT, TYPE, SVLEN, RECRUIT, DOSE) %>% mutate(COUNT = n()) %>% ungroup() %>% \n",
    "  distinct(POS, REF, ALT, TYPE, SVLEN, RECRUIT, DOSE, COUNT)\n",
    "t %>% filter(POS %in% c(89821199, 89821299)) %>% arrange(POS, ALT, desc(DOSE))\n",
    "\n",
    "snca %>% group_by(POS, PATNO) %>% filter(n()>1) # no variation for SVs for the same position in the same person\n",
    "\n",
    "snca_filt = snca %>% filter(POS %in% c(89821199, 89821299)) %>% \n",
    "  mutate(N_ALT_POS = paste(\"N_ALT\", POS, sep=\"_\")) %>% \n",
    "  tidyr::spread(key=N_ALT_POS, value=DOSE, fill = 0) %>% \n",
    "  group_by(PATNO) %>% \n",
    "  mutate(N_ALT_89821199=max(N_ALT_89821199),\n",
    "         N_ALT_89821299=max(N_ALT_89821299)) %>% ungroup() %>% \n",
    "  distinct(PATNO, .keep_all = T)\n",
    "\n",
    "# Create Analysis set\n",
    "full = dip %>% distinct(PATNO, .keep_all = T) %>% select(PATNO, RECRUIT) %>% \n",
    "  filter(RECRUIT %in% c(\"PD\", \"HC\"))\n",
    "\n",
    "df = snca_filt %>% select(-RECRUIT) %>% \n",
    "  left_join(full, ., by = \"PATNO\")\n",
    "df[is.na(df)]=0\n",
    "df %>% with(table(N_ALT_89821199, N_ALT_89821299))\n",
    "\n",
    "df %>% with(table(RECRUIT, N_ALT_89821199)) %>% chisq.test()\n",
    "df %>% with(table(RECRUIT, N_ALT_89821299)) %>% chisq.test()\n",
    "df %>% with(table(RECRUIT, I((N_ALT_89821199+(N_ALT_89821299))))) %>% chisq.test()\n",
    "\n",
    "pca = fread(\"../WGS/clean_NONGENT/3a.eigenvec\") %>% mutate(PATNO = sub(\"PPMISI\",\"\", IID) %>% as.numeric)\n",
    "df = inner_join(df, pca, by = \"PATNO\")\n",
    "glm(RECRUIT==\"PD\" ~ N_ALT_89821199+PC1+PC2+PC3+PC4+PC5+PC6+PC7+PC8+PC9+PC10, data = df, family = binomial()) %>% summary\n",
    "glm(RECRUIT==\"PD\" ~ I(N_ALT_89821299)+PC1+PC2+PC3+PC4+PC5+PC6+PC7+PC8+PC9+PC10, data = df, family = binomial()) %>% summary\n",
    "glm(RECRUIT==\"PD\" ~ I((N_ALT_89821199+(N_ALT_89821299>0))>0)+PC1+PC2+PC3+PC4+PC5+PC6+PC7+PC8+PC9+PC10, data = df, family = binomial()) %>% summary\n",
    "glm(RECRUIT==\"PD\" ~ I(N_ALT_89821199+N_ALT_89821299)+PC1+PC2+PC3+PC4+PC5+PC6+PC7+PC8+PC9+PC10, data = df, family = binomial()) %>% summary\n",
    "\"\"\"\n",
    "with open (\"script/exploreSNCA.R\", \"w\") as f:\n",
    "    f.write(t)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
